src/wallet/rpcwallet.cpp

```diff
@@ -3132,6 +3132,66 @@ UniValue generate(const JSONRPCRequest& request)
     return generateBlocks(coinbase_script, num_generate, max_tries, true);
 }
 
+UniValue sethdseed(const JSONRPCRequest& request)
+{
+    CWallet * const pwallet = GetWalletForJSONRPCRequest(request);
+
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp)) {
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() > 2) {
+        throw std::runtime_error(
+            "sethdseed ( \"flushkeypool\" \"seed\" )\n"
+            "\nSet or reset the HD wallet seed.\n"
+            + HelpRequiringPassphrase(pwallet) +
+            "\nArguments:\n"
+            "1. \"flushkeypool\"       (boolean, optional, default=true) Whether to flush old unused addresses from the keypool.\n"
+            "2. \"seed\"               (string, optional) A 64 character hex string to use as the new HD seed;\n"
+            "                             if not provided a random seed will be used.\n"
+            "\nResult:\n"
+            "true|false              (boolean) Returns true if successful\n"
+            "\nExamples:\n"
+            + HelpExampleCli("sethdseed", "")
+            + HelpExampleCli("sethdseed", "false")
+            + HelpExampleCli("sethdseed", "true \"c35c625bd2b3b20e1c98ed73d4dd2acf3c43a805d3fa303a6856ef1dd2a1e360\"")
+            + HelpExampleRpc("sethdseed", "true, \"c35c625bd2b3b20e1c98ed73d4dd2acf3c43a805d3fa303a6856ef1dd2a1e360\"")
+            );
+    }
+
+    LOCK2(cs_main, pwallet->cs_wallet);
+    EnsureWalletIsUnlocked(pwallet);
+
+    bool fFlushKeyPool = true;
+    if (request.params.size() > 0)
+        fFlushKeyPool = request.params[0].get_bool();
+
+    CPubKey masterPubKey;
+    if (request.params.size() > 1) {
+        CKey key;
+        std::string strKey = AccountFromValue(request.params[1]);
+        if (strKey.length() != 64 || !IsHex(strKey))
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Key should be 64 hex characters");
+
+        std::vector <unsigned char> vch = ParseHex(strKey);
+        key.Set(vch.begin(), vch.end(), true);
+
+        if (pwallet->HaveKey(key.GetPubKey().GetID()))
+            throw JSONRPCError(RPC_WALLET_ERROR, "Key already exists in wallet");
+
+        masterPubKey = pwallet->GenerateNewHDMasterKey(key);
+    } else
+        masterPubKey = pwallet->GenerateNewHDMasterKey();
```

Comment:
Yes, you would want to move funds from keys you suspect are compromised. The user is responsible for deciding how and when to move the funds. This pull request simply makes it possible to switch to a new uncompromised set of addresses.